Remove deprecations

connorjward opened the pull request:
## Description
* Removes the deprecated `.split()` method
* Removes a very old unused file
* Deprecates `V.subfunctions` in favour of `V.subspaces` (this needs some discussion)

## TODO

- [x] Make sure we don't raise our own warnings


<!--
Please include a summary of the changes introduced by this PR.
Additionally be sure to link associated pull requests in other projects.

If issues are fixed by this PR, include link to them and prepend each of them with the word "fixes", so they are automatically closed when this PR is merged.
For example "fixes #xyz, fixes #abc".

Feel free to add reviewers if you know there is someone who is already aware of this work.
Please open this PR initially as a draft and mark as ready for review once CI tests are passing.

Thanks for contributing!
-->

JHopeCollins commented:
LGTM, `.split` has been deprecated for a while now, good to have it removed.

I am very much in favour of changing `FunctionSpace.subfunctions` to `FunctionSpace.subspaces`. I haven't seen anywhere where being able to call `v.subfunctions` without knowing if `v` is a `Function` or `FunctionSpace` is useful enough to warrant the inaccurate naming.

connorjward commented:
> LGTM, `.split` has been deprecated for a while now, good to have it removed.
> 
> I am very much in favour of changing `FunctionSpace.subfunctions` to `FunctionSpace.subspaces`. I haven't seen anywhere where being able to call `v.subfunctions` without knowing if `v` is a `Function` or `FunctionSpace` is useful enough to warrant the inaccurate naming.
> 
> We should put a message in the slack channel warning that `Function.split` will finally be removed, say next Monday or something.

I have added this to [Wednesday's agenda](https://github.com/firedrakeproject/firedrake/wiki/Firedrake-meeting-2025-02-12). And assuming no issues we can merge later in the week following a couple of days warning.

JHopeCollins commented on firedrake/functionspaceimpl.py:
```diff
@@ -149,10 +149,17 @@ def topological(self, val):
+    @property
+    def subfunctions(self):
+        import warnings
+        warnings.warn("The 'subfunctions' property is deprecated, please use the "
```
```suggestion
        warnings.warn("The 'subfunctions' property is deprecated for 'FunctionSpace', please use the "
```

connorjward replied:
I tweaked the change a tiny bit, hope that's OK

JHopeCollins replied:
Sure, that's fine. It's just so that when people see it popping up they don't think they need to change all instances of `subfunctions`, only the ones for function spaces.

JHopeCollins commented on firedrake/functionspaceimpl.py:
```diff
@@ -641,14 +642,16 @@ def __str__(self):
         import warnings
-        warnings.warn("The .split() method is deprecated, please use the .subfunctions property instead", category=FutureWarning)
-        return self.subfunctions
+        warnings.warn("The 'subfunctions' property is deprecated, please use the "
```
```suggestion
        warnings.warn("The 'subfunctions' property is deprecated for 'FunctionSpace', please use the "
```

JHopeCollins commented on firedrake/functionspaceimpl.py:
```diff
@@ -1013,15 +1016,17 @@ def __hash__(self):
         import warnings
-        warnings.warn("The .split() method is deprecated, please use the .subfunctions property instead", category=FutureWarning)
-        return self.subfunctions
+        warnings.warn("The 'subfunctions' property is deprecated, please use the "
```
```suggestion
        warnings.warn("The 'subfunctions' property is deprecated for 'FunctionSpace', please use the "
```

JHopeCollins reviewed (changes requested):
I endorse this motion.
I've just left a suggestion to clarify the deprecation is only for `FunctionSpace`, not `Function`.

As we discussed offline, lets give a few days warning in slack before merging.
